### ISSUE TITLE
smc dropzone → _destroy → e.off is not a function

### ISSUE THREAD

**Issue by haraldschilly** [sagemathinc/smc#1875]
chrome: `Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/57.0.2987.133 Safari/537.36`

`e.off is not a function`

```
TypeError: e.off is not a function
    at Object._destroy (https://cloud.sagemath.com/static/smc-5d8110c797843a159773.js?5d8110c797843a159773:193:12377)
    at Object.componentWillUpdate (https://cloud.sagemath.com/static/smc-5d8110c797843a159773.js?5d8110c797843a159773:193:9643)
    at u._performComponentUpdate (https://cloud.sagemath.com/static/smc-5d8110c797843a159773.js?5d8110c797843a159773:187:5558)
    at updateComponent (https://cloud.sagemath.com/static/smc-5d8110c797843a159773.js?5d8110c797843a159773:187:4946)
    at u.receiveComponent (https://cloud.sagemath.com/static/smc-5d8110c797843a159773.js?5d8110c797843a159773:187:4057)
    at Object.receiveComponent (https://cloud.sagemath.com/static/smc-5d8110c797843a159773.js?5d8110c797843a159773:30:20190)
    at Object.updateChildren (https://cloud.sagemath.com/static/smc-5d8110c797843a159773.js?5d8110c797843a159773:186:30952)
    at h._reconcilerUpdateChildren (https://cloud.sagemath.com/static/smc-5d8110c797843a159773.js?5d8110c797843a159773:187:30032)
    at h._updateChildren (https://cloud.sagemath.com/static/smc-5d8110c797843a159773.js?5d8110c797843a159773:187:30797)
    at h.updateChildren (https://cloud.sagemath.com/static/smc-5d8110c797843a159773.js?5d8110c797843a159773:187:30695)
    at h._updateDOMChildren (https://cloud.sagemath.com/static/smc-5d8110c797843a159773.js?5d8110c797843a159773:187:17353)
```

top frame

```
this.dropzone.on(i,t);return this.dropzone.on("addedfile",function(e){return function(t){var n;if(t)return n=e.state.files,n.push(t),e.setState({files:n})}}(this))}},_destroy:function(e){if(null!=e)return e.off   [BANG]
  (),e.destroy()}})}
```

bottom of `smc-webapp/smc-dropzone.cjsx`

@johnjeng  Seems like this can't happen, right? Unless there is no `off` method in the first place. So far, I've only seen this once at `2017-04-19 11:02:55` ... so either a fluke or low probability.

time: 10 min

**Comment by williamstein** [comment]
@johnjeng: your code is:
```
        _destroy: (dropzone) ->
            return if not dropzone?
            dropzone.off()
            return dropzone.destroy()
```
Then the new @dropzone  is the output of `dropzone.destroy()`.   Why?????  Please add a comment explaining why to the code.   The actual dropzone.destroy() code is here:
https://github.com/enyo/dropzone/blob/master/src/dropzone.coffee#L730
It returns `Dropzone.instances.splice Dropzone.instances.indexOf(this), 1`, whatever that is.   The actual dropzone.destroy() function has no documentation except a line `# Not fully tested yet` before it...  In any case, explain what you were trying to do and why.  My guess is that you replace @dropzone by something random /undocument, which **is** defined, and which does not have an off method.

In short -- I think that this bug you've found @haraldschilly is exactly what I would expect from this code.  And this code needs to be documented way better.

**Comment by williamstein** [comment]
The SMC_Dropwrapper class starts with 8 space indenting, but our style in SMC is 4 spaces:
```
exports.SMC_Dropwrapper = rclass
        displayName: 'dropzone-wrapper'

        propTypes:
            project_id           : rtypes.string.isRequired    # The project to upload files to
            dest_path            : rtypes.string.isRequired    # The path for files to be sent
```

I don't know how that happened -- maybe I messed it up when copying it out of r_misc; please fix when fixing this bug..
